plugins/ccgx/fu-ccgx-native-hid-device.c

```diff
@@ -0,0 +1,505 @@
+			 FU_CCGX_PD_RESP_BRIDGE_MODE_CMD_SIG,
+			 0x43,
+			 0x59,
+			 0x00,
+			 0x00,
+			 0x00,
+			 0x0B};
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(self),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_NATIVE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error)) {
+		g_prefix_error(error, "magic enable command error: ");
+		return FALSE;
+	}
+
+	/* ignore error: this always fails but has the correct behavior */
+	if (!fu_ccgx_native_hid_command(self,
+					CCGX_HID_CMD_MODE,
+					FU_CCGX_PD_RESP_BRIDGE_MODE_CMD_SIG,
+					&error_local)) {
+		g_debug("expected HID report bridge mode failure: %s", error_local->message);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_native_hid_ensure_fw_info(FuCcgxNativeHidDevice *self, GError **error)
+{
+	guint8 buf[0x40] = {FU_CCGX_NATIVE_HID_REPORT_ID_INFO, 0};
+	g_autofree gchar *bl_ver = NULL;
+	g_autoptr(GByteArray) st_info = NULL;
+
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+				      buf[0],
+				      buf,
+				      sizeof(buf),
+				      FU_CCGX_NATIVE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error))
+		return FALSE;
+	if (!fu_ccgx_native_hid_enter_flashing_mode(self, error))
+		return FALSE;
+
+	st_info = fu_struct_ccgx_native_hid_fw_info_parse(buf, sizeof(buf), 0x0, error);
+	if (st_info == NULL)
+		return FALSE;
+	self->silicon_id = fu_struct_ccgx_native_hid_fw_info_get_silicon_id(st_info);
+	self->fw_mode = fu_struct_ccgx_native_hid_fw_info_get_operating_mode(st_info);
+
+	/* set current version
+	 * Note that this is the base version, not the app version.
+	 */
+	if (self->fw_mode == FU_CCGX_FW_MODE_FW1) {
+		guint version = fu_struct_ccgx_native_hid_fw_info_get_image1_base_version(st_info);
+		fu_device_set_version_from_uint32(FU_DEVICE(self), version);
+	} else if (self->fw_mode == FU_CCGX_FW_MODE_FW2) {
+		guint version = fu_struct_ccgx_native_hid_fw_info_get_image2_base_version(st_info);
+		fu_device_set_version_from_uint32(FU_DEVICE(self), version);
+	}
+
+	/* set bootloader version */
+	fu_device_set_version_bootloader_raw(
+	    FU_DEVICE(self),
+	    fu_struct_ccgx_native_hid_fw_info_get_bl_version(st_info));
+	bl_ver = fu_version_from_uint32(fu_struct_ccgx_native_hid_fw_info_get_bl_version(st_info),
+					fu_device_get_version_format(self));
+	fu_device_set_version_bootloader(FU_DEVICE(self), bl_ver);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_native_hid_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	fu_device_remove_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	fu_device_remove_private_flag(device, FU_CCGX_NATIVE_HID_DEVICE_IS_IN_RESTART);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_native_hid_device_close(FuDevice *device, GError **error)
+{
+	/* do not close handle when device restarts */
+	if (fu_device_has_private_flag(device, FU_CCGX_NATIVE_HID_DEVICE_IS_IN_RESTART))
+		return TRUE;
```

Comment:
do we need this? i think the daemon just ignores the error if we close a device that doesn't exist.